### PR TITLE
Placeholder - user support for past credit

### PR DESCRIPTION
**Pull request by tskeeley**
Updates the deeplinks process and allows for requesting attendance at past meetings.

NOTE:  This makes a record of the request ONLY.  It does not yet send the request for approval as, I am uncomfortable with making that change at the same time.

This patch DOES through PHP notices about missing index, as the name lookup seems to fail, but I need that for the "part 2" of this.

Please do a run of this in your working environment of both a past meeting AND signin-today meeting so we know it's not going to break things in prod

### CODE REVIEW

File: data/database-schema.php [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 class SCHEMA
 {
-  public static $REQUIED_DB_SCHEMA = 2018092400; // Current DB Version - YYYYMMDDvv format (vv=daily counter form 00)
+  public static $REQUIED_DB_SCHEMA = 2019111500; // Current DB Version - YYYYMMDDvv format (vv=daily counter form 00)
```
[the remark this answers is not in the document]
Comment (AricStewart) [reply data/database-schema.php]: This file is presently phpcs non-compliant and that is ok for now.

File: data/database-schema.php [hunk 2]
```diff
@@ -51,6 +51,16 @@ class SCHEMA
         'RegistrationID' => 'INT UNSIGNED NOT NULL PRIMARY KEY',  // 1:1 mapping of the Registrations Primary Key
         'Page' => 'INT UNSIGNED NOT NULL',
     ],
+    'DeepLinks' => [
+        'LinkID' => 'INT UNSIGNED NOT NULL PRIMARY KEY AUTO_INCREMENT',
```
[the remark this answers is not in the document]
Comment (AricStewart) [reply data/database-schema.php]: This file is presently phpcs non-compliant and that is ok for now.

File: data/database-schema.php [hunk 3]
```diff
@@ -51,6 +51,16 @@ class SCHEMA
...
+        'AccountID' => 'INT UNSIGNED NOT NULL',
```
[the remark this answers is not in the document]
Comment (AricStewart) [reply data/database-schema.php]: This file is presently phpcs non-compliant and that is ok for now.

File: data/database-schema.php [hunk 4]
```diff
@@ -51,6 +51,16 @@ class SCHEMA
...
+        'ApproverID' => 'INT UNSIGNED NOT NULL',
```
[the remark this answers is not in the document]
Comment (AricStewart) [reply data/database-schema.php]: This file is presently phpcs non-compliant and that is ok for now.

File: data/database-schema.php [hunk 5]
```diff
@@ -51,6 +51,16 @@ class SCHEMA
...
+        'ApprovedValue' => 'TEXT NOT NULL',
+        'Auth' => 'VARCHAR(40) NOT NULL',
```
[the remark this answers is not in the document]
Comment (AricStewart) [reply data/database-schema.php]: This file is presently phpcs non-compliant and that is ok for now.

File: data/database-schema.php [hunk 6]
```diff
@@ -51,6 +51,16 @@ class SCHEMA
...
+        'Date' => 'TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP',
```
[the remark this answers is not in the document]
Comment (AricStewart) [reply data/database-schema.php]: This file is presently phpcs non-compliant and that is ok for now.

File: data/database-schema.php [hunk 7]
```diff
@@ -51,6 +51,16 @@ class SCHEMA
...
+        'Description' => 'VARCHAR(100) NOT NULL',
```
[the remark this answers is not in the document]
Comment (AricStewart) [reply data/database-schema.php]: This file is presently phpcs non-compliant and that is ok for now.

File: data/database-schema.php [hunk 8]
```diff
@@ -51,6 +51,16 @@ class SCHEMA
...
+        'Type' => 'VARCHAR(10) NOT NULL',
```
[the remark this answers is not in the document]
Comment (AricStewart) [reply data/database-schema.php]: This file is presently phpcs non-compliant and that is ok for now.

File: data/database-schema.php [hunk 9]
```diff
@@ -51,6 +51,16 @@ class SCHEMA
         'RegistrationID' => 'INT UNSIGNED NOT NULL PRIMARY KEY',  // 1:1 mapping of the Registrations Primary Key
         'Page' => 'INT UNSIGNED NOT NULL',
     ],
+    'DeepLinks' => [
+        'LinkID' => 'INT UNSIGNED NOT NULL PRIMARY KEY AUTO_INCREMENT',
+        'AccountID' => 'INT UNSIGNED NOT NULL',
+        'ApproverID' => 'INT UNSIGNED NOT NULL',
+        'ApprovedValue' => 'TEXT NOT NULL',
```
[the remark this answers is not in the document]
Comment (AricStewart) [reply data/database-schema.php]: This file is presently phpcs non-compliant and that is ok for now.

**Comment by AricStewart** [comment]
DeepLinks? why that name?
Some spacing on your strings need adjusting.
I would REALLY prefer a button to a link for the request.

**Comment by mshappe** [comment]
Also, I'm curious why we are solving a problem I'm not sure we have. We
don't currently vet LIVE attendance checkins. Why would we vet past ones?

I mean, really, anyone could sign in and flag themselves attending today,
then turn over and go back to sleep. Nobody is ever going to check,
especially now that remote attendance is a thing.

**Comment by tskeeley** [comment]
@AricStewart I have no idea how to clear the spacing/quality checks on these, since they are old an my changes followed that format.  I'm not asking you to do, I am asking to learn.

I was hoping that this could go out today before the meeting, but since it didn't, I think we could step back and look at the rest of the process.

@mshappe The reason was because the "board" asked for the right to approve.  In this case, the "board" is represented by Lauren, Amy, Liz.  I don't know it others agreed or not.  But the direction I was given follows the thought that anyone can sign in on the day of a meeting, because, they could be online or not.  It is a limited period of time (used to be JUST during the meeting, but now it is anytime that day).  The 'past credit' is anytime at all, and the thought is that it would be abused by people "clicking away" (intentionally or not) and there was a concern that it would invalidate the legal attendance voting requirements of the bylaws in terms of "willful neglect".  Also, it opened up the abuse for fake coo (bring a bunch of people to vote, have them all claim they were at meetings, and presto, you have a bunch of unqualified people taking over the elections in a "legal" way.

SO... I'm willing to undo the approval requirement, but we need to be very sure we are accepting that issue which is WAY above my pay grade, even when I was on the board. :)

File: modules/concom/pages/panes.inc [hunk 10]
```diff
@@ -105,7 +105,17 @@ function meetings()
...
-                    echo "<td>Not Checked In";
+                    echo "<td>Not Checked In -";
+                    $checkPending = check_meeting_pastrequest($value['MeetingID']);
+
+                    if ($checkPending) {
+                        echo "Attendance request pending";
+                    } else {
+                      echo "<form action='index.php?Function=concom/pastcredit&meeting=".$value['MeetingID']."' method='post' name='butiwasthere' id='butiwasthere' class='UI-container'>";
+                      echo "    <input type='hidden' name='meeting' value='".$value['MeetingID']."'>";
+                      echo "    <button class='UI-OrangeButton'>But I attended!</button>";
```
Comment (AricStewart) [inline modules/concom/pages/panes.inc]: UI-OrangeButton is not defined anywhere as a style. you want "UI-orangebutton"

File: modules/concom/pages/panes.inc [hunk 11]
```diff
@@ -105,7 +105,17 @@ function meetings()
                 if ($nowDate == $checkDate) {
                     echo "<td class='CONCOM-checkin-now'><a href='index.php?Function=concom/signin&meeting=".$value['MeetingID']."'>Sign In Now!</a>";
                 } else {
-                    echo "<td>Not Checked In";
+                    echo "<td>Not Checked In -";
+                    $checkPending = check_meeting_pastrequest($value['MeetingID']);
+
+                    if ($checkPending) {
+                        echo "Attendance request pending";
```
Comment (AricStewart) [inline modules/concom/pages/panes.inc]: Would we please get a space at the beginning of this string.  "Not Checked In -Attendance request pending"  no space between the '-' and 'A' bugs me.

**Comment by AricStewart** [comment]
My phpcs freaks out about a lot of these changes. I do not know why the one in the actions is not...

```
FILE: ...sers/aric/Personal/CON/CIAB-Signin/modules/concom/functions/ATTENDANCE.inc
--------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------
 9 | ERROR | [x] Expected 2 blank lines before function; 1 found
--------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
--------------------------------------------------------------------------------


FILE: /Users/aric/Personal/CON/CIAB-Signin/modules/concom/pages/panes.inc
--------------------------------------------------------------------------------
FOUND 5 ERRORS AFFECTING 5 LINES
--------------------------------------------------------------------------------
 114 | ERROR | [x] Line indented incorrectly; expected 24 spaces, found 22
 115 | ERROR | [x] Line indented incorrectly; expected 24 spaces, found 22
 116 | ERROR | [x] Line indented incorrectly; expected 24 spaces, found 22
 117 | ERROR | [x] Line indented incorrectly; expected 24 spaces, found 22
 127 | ERROR | [x] Expected 1 blank line before closing function brace; 0 found
--------------------------------------------------------------------------------
PHPCBF CAN FIX THE 5 MARKED SNIFF VIOLATIONS AUTOMATICALLY
--------------------------------------------------------------------------------

Time: 1.44 secs; Memory: 12Mb
```

File: modules/concom/pages/panes.inc (hunk 11 above)
Comment (tskeeley) [reply modules/concom/pages/panes.inc]: Got it

File: modules/concom/pages/panes.inc (hunk 10 above)
Comment (tskeeley) [reply modules/concom/pages/panes.inc]: I couldn't see it, thank you

**Pull request closed without merging**